lib/optimizer/tests/SpecTest.php

```diff
@@ -108,33 +108,99 @@ public function testTransformerSpecFiles($scenario, $transformerClass, $source,
         }
 
+        $configuration = $this->mapConfigurationData($this->extractConfigurationData($source));
+
         $document = Document::fromHtmlFragment($source);
 
-        $transformer = $this->getTransformer($scenario, $transformerClass);
-        $errors      = new ErrorCollection();
+        $transformer   = $this->getTransformer($scenario, $transformerClass, $configuration);
+        $errors        = new ErrorCollection();
 
         $transformer->transform($document, $errors);
 
         $this->assertSimilarMarkup($expected, $document->saveHTMLFragment());
     }
 
+    /**
+     * Map spec test input file configuration data to configuration arguments as needed by the PHP transformers.
+     *
+     * @param array $configurationData Associative array of configuration data coming from the spec test input file.
+     * @return Configuration Configuration object to use for the transformation engine.
+     */
+    public function mapConfigurationData($configurationData)
+    {
+        $mappedConfiguration = [];
+
+        foreach ($configurationData as $key => $value) {
+            switch ($key) {
+                case 'ampRuntimeStyles':
+                    $mappedConfiguration[AmpRuntimeCss::class][AmpRuntimeCssConfiguration::STYLES] = $value;
+                    break;
+                case 'ampRuntimeVersion':
+                    $mappedConfiguration[AmpRuntimeCss::class][AmpRuntimeCssConfiguration::VERSION] = $value;
+                    break;
+
+                // @TODO: Known configuration arguments used in spec tests that are not implemented yet.
+                case 'ampUrlPrefix':
+                case 'ampUrl':
+                case 'canonical':
+                case 'experimentBindAttribute':
+                case 'geoApiUrl':
+                case 'lts':
+                case 'preloadHeroImage':
+                case 'rtv':
+                default:
+                    $this->fail("Configuration argument not yet implemented: {$key}.");
+            }
+        }
+
+        return new Configuration($mappedConfiguration);
+    }
+
+    /**
+     * Parse the input source file and extract the configuration data.
+     *
+     * Input HTML files can contain a leading HTML comment that provides configuration arguments in the form of a JSON
+     * object.
+     *
+     * @param string $source Input source file to parse for a configuration snippet.
+     * @return array Associative array of configuration data found in the input HTML file.
+     */
+    public function extractConfigurationData(&$source)
+    {
+        $matches = [];
+        if (!preg_match(self::LEADING_HTML_COMMENT_REGEX_PATTERN, $source, $matches)) {
+            return [];
+        }
+
+        $json   = trim($matches['json']);
+        $source = str_replace($matches[0], '', $source);
+
+        if (empty($json)) {
+            return [];
+        }
+
+        $configurationData = (array)json_decode($json, true, 512, JSON_OBJECT_AS_ARRAY);
```

Comment:
Why is `JSON_OBJECT_AS_ARRAY` needed if `assoc=true` is the second argument?

Reply:
Resolved via https://github.com/ampproject/amp-wp/pull/4662/commits/96c7991cfe4808f97d87623b7c07e6c1d03df10c